File: src/dynamixel/conversions.py
"""Unit conversion functions for Dynamixel motors."""

import math
from typing import Dict

from .constants import POSITION_RESOLUTION, CENTER_POSITION, JointConfig


def ticks_to_radians(ticks: int) -> float:
    """Convert Dynamixel position ticks to radians.

    Dynamixel XM series motors use 4096 ticks per revolution.
    Position 2048 is center (0 radians).
    Range: 0-4095 maps to approximately -pi to +pi.

    Args:
        ticks: Raw position value from motor (0-4095 for single turn)

    Returns:
        Position in radians
    """
    offset = ticks - CENTER_POSITION
    radians = (offset / POSITION_RESOLUTION) * 2 * math.pi
    return radians


def radians_to_ticks(radians: float) -> int:
    """Convert radians to Dynamixel position ticks.

    Args:
        radians: Position in radians

    Returns:
        Position ticks for motor command
    """
    offset = (radians / (2 * math.pi)) * POSITION_RESOLUTION
    ticks = int(round(CENTER_POSITION + offset))
    return ticks


def radians_to_degrees(radians: float) -> float:
    """Convert radians to degrees."""
    return math.degrees(radians)


def degrees_to_radians(degrees: float) -> float:
    """Convert degrees to radians."""
    return math.radians(degrees)


def validate_joint_limits(
    joint_index: int, radians: float, joints_config: Dict[int, JointConfig]
) -> bool:
    """Check if position is within joint limits.

    Args:
        joint_index: Index of the joint (0-5)
        radians: Target position in radians
        joints_config: Joint configuration dictionary

    Returns:
        True if within limits, False otherwise
    """
    if joint_index not in joints_config:
        return False

    config = joints_config[joint_index]
    degrees = radians_to_degrees(radians)
    return config.min_degrees <= degrees <= config.max_degrees

File: dynamixel/conversions.go
package dynamixel

import (
	"encoding/binary"
	"math"
)

// TicksToRadians converts Dynamixel encoder ticks to radians.
// The motor has 4096 ticks per revolution with center at 2048.
func TicksToRadians(ticks int) float64 {
	// Offset from center position, then convert to radians
	offset := float64(ticks - CenterPosition)
	return offset * (2.0 * math.Pi / float64(TicksPerRevolution))
}

// RadiansToTicks converts radians to Dynamixel encoder ticks.
func RadiansToTicks(radians float64) int {
	offset := radians * (float64(TicksPerRevolution) / (2.0 * math.Pi))
	return int(math.Round(offset)) + CenterPosition
}

// RadiansToDegrees converts radians to degrees.
func RadiansToDegrees(radians float64) float64 {
	return radians * 180.0 / math.Pi
}

// ValidateJointLimits checks if a position is within joint limits.
func ValidateJointLimits(jointIdx int, radians float64) bool {
	if jointIdx < 0 || jointIdx >= len(ViperX300SJoints) {
		return false
	}
	config := ViperX300SJoints[jointIdx]
	return radians >= config.MinRadians && radians <= config.MaxRadians
}

// BytesToInt32 converts 4 bytes (little-endian) to an int32.
func BytesToInt32(data []byte) int32 {
	if len(data) < 4 {
		return 0
	}
	return int32(binary.LittleEndian.Uint32(data))
}

// Int32ToBytes converts an int32 to 4 bytes (little-endian).
func Int32ToBytes(val int32) []byte {
	buf := make([]byte, 4)
	binary.LittleEndian.PutUint32(buf, uint32(val))
	return buf
}
